docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+### Rationale
+
+There exists a desire for incremental source generators be able to write out a set of accompanying files, alongside the generated sources that they add to the compilation. An example of this is JavaScript interop with .NET code. When exposing a .NET method to JavaScript today it is possible to generate the C# interop source code that enables the call to be made from JavaScript, but the user must still hand author the corresponding JavaScript code. Instead, with this feature the incremental generator could emit both the required C# code alongside a JavaScript file that provides the equivalent JavaScript interop code.
+
+### High Level Design Goals
+
+- Allow Incremental Generators to emit source or binary files to disk alongside the emitted .NET assembly
+- The user can control where the generated artifacts are emitted with a suitable default
+- Be optimal in terms of file writes / copies and ensure we keep the correct incremental build semantics for MSBuild
+- Do *not* run as part of the IDE experience. Files are only emitted to disk as part of command line compilation
+
+## Implementation
+
+The feature is exposed to incremental generator authors via a new output API:
+
+```csharp
+namespace Microsoft.CodeAnalysis;
+
+public readonly partial struct IncrementalGeneratorInitializationContext
+{
+    public void RegisterArtifactOutput<TSource>(IncrementalValueProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+    public void RegisterArtifactOutput<TSource>(IncrementalValuesProvider<TSource> source, Action<ArtifactProductionContext, TSource> action);
+}
+
+public readonly struct ArtifactProductionContext
+{
+    public CancellationToken CancellationToken { get; }
+
+    public void AddFile(string hintName, Action<System.IO.Stream> callback);
+}
+```
+
+This can be used by a generator author like so:
+
+```csharp
+internal sealed class Generator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var compilationSource = context.CompilationProvider;
+
+        context.RegisterArtifactOutput(compilationSource, (nspc, item) =>
+        {
+            nspc.AddFile("interop.js", (stream) =>
+            {
+                using var writer = new StreamWriter(stream);
+                writer.WriteLine("//output");
+            });
+        });
+    }
+}
+```
+
+Note how the call to `AddFile` takes a callback that provides a stream. This allows the host of the generator driver to control how the stream is created and what kind of stream is created. For instance in the compiler it could be a `FileStream` that writes directly to disk, but in unit tests it could be a `MemoryStream` that can be inspected as part of the test. It also takes a `hintName`, the same way that source files do, rather than an explicit file name or path. This is explained in more detail below.
```

Comment:
Hmm, this is an interesting point. I had assumed we would just write the file essentially at the point that it's called in the generator, but I agree that we probably shouldn't be stamping on stuff in the case of a failed build. 

We could certainly hold onto the callback and not invoke it until the point we're actually ready to write it. That brings up some possible weirdness around closures and lifetimes of things in the generator, although in theory it should only be the item passed into the outer `AddArtifactOutput` which is coming from a state table and thus hanging around anyway.

Then there is the alternative take which says, oh, maybe we should just be writing into the `obj` directory and copying upon success.

Reply:
Hmm, actually the `obj` thing is interesting to me, because we could still combine it with the idea of `<CompilerArtifactMapping` but we'd just write all the files into dedicated `obj` directories and have the compiler targets do the copying based on the mappings. 

That's actually much simpler, achieves the same thing from a user and author pov. The only downside is the extra copy that isn't present by writing it directly. We'd also need a lot less  compiler infrastructure as we'd just write it to the existing `generatedFilesOut` directory with a dedicated artifacts subfolder.

Reply:
If we combined the allowing of subfolders too, then the folder could stand in for the 'virtual' folder name. You could imagine the mapping actually being something like

Copy all artifacts and subfolders to a specific folder:
`<CompilerArtifactMapping Include="GeneratorName" OutputDir="$(MSBuildProjectRoot)\wwwroot" />`

Copy specific parts to other places
```
<CompilerArtifactMapping Include="GeneratorName" SourceDir="js" OutputDir="$(MSBuildProjectRoot)\wwwroot\js" />
<CompilerArtifactMapping Include="GeneratorName" SourceDir="js\interop" OutputDir="$(MSBuildProjectRoot)\wwwroot\specialInteropStuff" />
```

Reply:
> I had assumed we would just write the file essentially at the point that it's called in the generator, but I agree that we probably shouldn't be stamping on stuff in the case of a failed build.

If we move to the model of being able to see other generators output then this becomes maybe a bit easier. That could be structured to run after core compile completes so by then we know that compilation succeeded.